File: xamarin/Callibri/CallibriDemo/CallibriDemo/NeuroImpl/CallibriController.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;

using CallibriDemo.Utils;

using NeuroSDK;

using Xamarin.Forms;

namespace CallibriDemo.NeuroImpl;

internal class CallibriController
{
#region Info
    public string GetInfo()
    {
        if (_sensor?.State == SensorState.StateOutOfRange) return "Device unreachable!";

        var deviceInfo = "";

        var features = _sensor.Features;
        deviceInfo += "\n===== [Features] =====\n";
        deviceInfo =  features.Aggregate(deviceInfo, (current, feature) => current + $"[{feature}]: {_sensor.IsSupportedFeature(feature)}\n");

        var commands = _sensor.Commands;
        deviceInfo += "\n===== [Commands] =====\n";
        deviceInfo =  commands.Aggregate(deviceInfo, (current, command) => current + $" {command}\n");

        var parameters = _sensor.Parameters;
        deviceInfo += "\n===== [Parameters] =====\n";
        foreach (ParameterInfo parameter in parameters)
        {
            switch (parameter.Param)
            {
            case SensorParameter.ParameterName:
                deviceInfo += $" Name: {_sensor.Name} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterState:
                deviceInfo += $" Connection state: {_sensor.State} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterAddress:
                deviceInfo += $" Address: {_sensor.Address} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSerialNumber:
                deviceInfo += $" Serial number: {_sensor.SerialNumber} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterHardwareFilterState:
                deviceInfo += $" Hardware filters (Access: {parameter.ParamAccess}): \n";
                deviceInfo =  _sensor.HardwareFilters.Aggregate(deviceInfo, (current, filter) => current + $" {filter}\n");
                break;
            case SensorParameter.ParameterFirmwareMode:
                deviceInfo += $" Firmware mode: {_sensor.FirmwareMode} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSamplingFrequency:
                deviceInfo += $" Sampling frequency: {_sensor.SamplingFrequency} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterGain:
                deviceInfo += $" Gain: {_sensor.Gain} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterOffset:
                deviceInfo += $" Offset: {_sensor.DataOffset} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterExternalSwitchState:
                deviceInfo += $" External switch state: {_sensor.ExtSwInput} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterADCInputState:
                deviceInfo += $" ADCInput state: {_sensor.ADCInput} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterAccelerometerSens:
                deviceInfo += $" Accelerometer sensitivity: {_sensor.AccSens} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterGyroscopeSens:
                deviceInfo += $" Gyroscope sensitivity: {_sensor.GyroSens} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterStimulatorAndMAState:
                deviceInfo += $" Stimulator state: {_sensor.StimulatorMAStateCallibri.StimulatorState} (Access: {parameter.ParamAccess})\n";
                deviceInfo += $" Stimulator state: {_sensor.StimulatorMAStateCallibri.MAState} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterStimulatorParamPack:
                deviceInfo += $" Stimulator param pack (Access: {parameter.ParamAccess}):\n";
                deviceInfo += $"  PulseWidth: {_sensor.StimulatorParamCallibri.PulseWidth}\n";
                deviceInfo += $"  StimulusDuration: {_sensor.StimulatorParamCallibri.StimulusDuration}\n";
                deviceInfo += $"  Frequency: {_sensor.StimulatorParamCallibri.Frequency}\n";
                deviceInfo += $"  Current: {_sensor.StimulatorParamCallibri.Current}\n";
                break;
            case SensorParameter.ParameterMotionAssistantParamPack:
                deviceInfo += $" MotionAssistantParamPack (Access: {parameter.ParamAccess}):\n";
                deviceInfo += $"  GyroStop: {_sensor.MotionAssistantParamCallibri.GyroStop}\n";
                deviceInfo += $"  GyroStart: {_sensor.MotionAssistantParamCallibri.GyroStart}\n";
                deviceInfo += $"  Limb: {_sensor.MotionAssistantParamCallibri.Limb}\n";
                deviceInfo += $"  MinPauseMs: {_sensor.MotionAssistantParamCallibri.MinPauseMs}\n";
                break;
            case SensorParameter.ParameterFirmwareVersion:
                deviceInfo += $" Sensor version (Access: {parameter.ParamAccess}): "
                            + $"[FW]: {_sensor.Version.FwMajor}.{_sensor.Version.FwMinor} "
                            + $"[HW]: {_sensor.Version.HwMajor}.{_sensor.Version.HwMinor}.{_sensor.Version.HwPatch} "
                            + $"[Ext]: {_sensor.Version.ExtMajor}\n";
                break;
            case SensorParameter.ParameterMEMSCalibrationStatus:
                deviceInfo += $" {_sensor.MEMSCalibrateStateCallibri}(Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterMotionCounterParamPack:
                deviceInfo += $" MotionCounterParamPack (Access: {parameter.ParamAccess}):\n";
                deviceInfo += $"  InsenseThresholdMG: {_sensor.MotionCounterParamCallibri.InsenseThresholdMG}\n";
                deviceInfo += $"  InsenseThresholdSample: {_sensor.MotionCounterParamCallibri.InsenseThresholdSample}\n";
                break;
            case SensorParameter.ParameterMotionCounter:
                deviceInfo += $" Motion counter (Access: {parameter.ParamAccess}): {_sensor.MotionCounterCallibri}\n";
                break;
            case SensorParameter.ParameterBattPower:
                deviceInfo += $" Battery: {_sensor.BattPower} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSensorFamily:
                deviceInfo += $"Sensor family: {_sensor.SensFamily} (Access: {parameter.ParamAccess})\n";
                break;
            case SensorParameter.ParameterSensorMode:
                deviceInfo += $" Sensor mode (Access: {parameter.ParamAccess}): {_sensor.FirmwareMode}\n";
                break;
            case SensorParameter.ParameterSamplingFrequencyResp:
                deviceInfo += $" (Access: {parameter.ParamAccess})\n";
                break;
            }
        }

        deviceInfo += $" Signal type: {_sensor.SignalTypeCallibri}\n";

        deviceInfo += $" ColorCallibri: {_sensor.ColorCallibri}\n";

        return deviceInfo;
    }
#endregion

#region Singleton
    private static CallibriController _instance;

    public static CallibriController Instance { get => _instance ??= new CallibriController(); }
#endregion

#region Search
    private Scanner        _scanner;
    private SensorsChanged _sensorFounded;

    public void StartSearch(SensorsChanged sensorChanged)
    {
        ISensorHelper sensorHelper = DependencyService.Get<ISensorHelper>();

        sensorHelper.EnableSensor(
            enabled =>
            {
                if (!enabled) return;

                _scanner ??= new Scanner(SensorFamily.SensorLECallibri, SensorFamily.SensorLEKolibri);

                _sensorFounded               =  sensorChanged;
                _scanner.EventSensorsChanged += _sensorFounded;
                _scanner?.Start();
            }
        );
    }

    public void StopSearch()
    {
        _scanner?.Stop();
        if (_scanner != null) _scanner.EventSensorsChanged -= _sensorFounded;
        _sensorFounded = null;
    }
#endregion

#region Sensor
    private CallibriSensor _sensor;

    public SensorStateChanged ConnectionStateChanged;
    public BatteryChanged     BatteryChanged;

    public bool IsEnvelopeSupported
    {
        get
        {
            try
            {
                if (_sensor?.State == SensorState.StateInRange)
                {
                    return _sensor.IsSupportedCommand(SensorCommand.CommandStartEnvelope);
                }
                else return false;
            }
            catch (Exception e)
            {
                Debug.WriteLine(e);
            }

            return false;
        }
    }

    public SensorState ConnectionState { get => _sensor?.State ?? SensorState.StateOutOfRange; }

    public int BatteryPower { get => _sensor?.BattPower ?? 0; }

    public int SamplingFrequency
    {
        get
        {
            try
            {
                SensorSamplingFrequency samplingFrequency = _sensor?.SamplingFrequency ?? 0;
                switch (samplingFrequency)
                {
                case SensorSamplingFrequency.FrequencyHz10:
                    return 10;
                case SensorSamplingFrequency.FrequencyHz20:
                    return 20;
                case SensorSamplingFrequency.FrequencyHz100:
                    return 100;
                case SensorSamplingFrequency.FrequencyHz125:
                    return 125;
                case SensorSamplingFrequency.FrequencyHz250:
                    return 250;
                case SensorSamplingFrequency.FrequencyHz500:
                    return 500;
                case SensorSamplingFrequency.FrequencyHz1000:
                    return 1000;
                case SensorSamplingFrequency.FrequencyHz2000:
                    return 2000;
                case SensorSamplingFrequency.FrequencyHz4000:
                    return 4000;
                case SensorSamplingFrequency.FrequencyHz8000:
                    return 8000;
                case SensorSamplingFrequency.FrequencyUnsupported:
                default:
                    return 0;
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine(e);
            }

            return 0;
        }
    }

    public Task<SensorState> CreateAndConnect(SensorInfo info)
    {
        SensorState outState = SensorState.StateOutOfRange;
        return Task<SensorState>.Factory.StartNew(
            () =>
            {
                try
                {
                    _sensor = _scanner.CreateSensor(info) as CallibriSensor;
                    if (_sensor != null)
                    {
                        _sensor.EventSensorStateChanged += ConnectionStateChanged;
                        _sensor.EventBatteryChanged     += BatteryChanged;

                        _sensor.SamplingFrequency = SensorSamplingFrequency.FrequencyHz1000;

                        _sensor.SignalTypeCallibri = CallibriSignalType.ECG;
                        outState = _sensor.State;
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.ToString());
                }
                return outState;
            }
        );
    }

    public Task<SensorState> ConnectCurrent()
    {
        return Task<SensorState>.Factory.StartNew(
            () =>
            {
                if (_sensor.State != SensorState.StateOutOfRange) return SensorState.StateInRange;

                _sensor?.Connect();
                return _sensor.State;
            }
        );
    }

    public void DisconnectCurrent()
    {
        try
        {
            _sensor?.Disconnect();
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.ToString());
        }
    }
#endregion

#region Signal
    public CallibriSignalDataRecived   SignalReceived;
    public CallibriEnvelopeDataRecived EnvelopeReceived;

    public void StartSignal()
    {
        _sensor.EventCallibriSignalDataRecived += SignalReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStartSignal);
    }

    public void StopSignal()
    {
        _sensor.EventCallibriSignalDataRecived -= SignalReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStopSignal);
    }

    public void StartEnvelope()
    {
        _sensor.EventCallibriEnvelopeDataRecived += EnvelopeReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStartEnvelope);
    }

    public void StopEnvelope()
    {
        _sensor.EventCallibriEnvelopeDataRecived -= EnvelopeReceived;
        _sensor?.ExecCommand(SensorCommand.CommandStopEnvelope);
    }
#endregion
}

File: react-native/callibri/NeuroImpl/CallibriController.ts
import { Scanner, 
  SensorInfo, 
  SensorFamily, 
  CallibriSensor, 
  SensorState, 
  SensorCommand, 
  SensorFeature, 
  SensorParameter, 
  SensorParamAccess, 
  SensorFirmwareMode, 
  SensorDataOffset, 
  SensorGain, 
  SensorSamplingFrequency, 
  CallibriSignalData, 
  CallibriEnvelopeData, 
  CallibriElectrodeState, 
  CallibriSignalType, 
  CallibriColorType, 
  SensorFilter, 
  SensorADCInput, 
  SensorGyroscopeSensitivity, 
  SensorAccelerometerSensitivity, 
  SensorExternalSwitchInput } from "react-native-neurosdk2";
import { EventSubscription, PermissionsAndroid, Platform } from 'react-native';

let instance: CallibriController;

class CallibriController {

  private static _instance: CallibriController;

  private constructor() {
    //...
  }

  public static get Instance() {
    return this._instance || (this._instance = new this());
  }

  private _scanner: Scanner | undefined

  private async requestPermissionAndroid() {
    try {
      const result = await PermissionsAndroid.requestMultiple([
        PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
        PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT,
      ]);
      if (
        result[PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION] !=
        PermissionsAndroid.RESULTS.GRANTED
      ) {
        this.requestPermissionAndroid();
      }
      if (
        result[PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT] !=
        PermissionsAndroid.RESULTS.GRANTED
      ) {
        this.requestPermissionAndroid();
      }
    } catch (err) {
      console.warn(err);
    }
  }

  async startSearch(sensorFounded: (sensorFounded: SensorInfo[]) => void): Promise<void> {

    if (this._scanner != undefined) {
      this._scanner.close()
      this._scanner = undefined
    }
    this._scanner = new Scanner()
    await this._scanner.init([SensorFamily.LECallibri, SensorFamily.LEKolibri])
    this._scanner?.AddSensorListChanged(sensorFounded)
    if(Platform.OS != 'ios'){
      await this.requestPermissionAndroid()
    }
    await this._scanner?.start()
  }

  async stopSearch(): Promise<void> {
    this._scanner?.RemoveSensorListChanged()
    await this._scanner?.stop()
  }

  private _sensor: CallibriSensor | undefined
  public connectionChangedCallback: ((state: SensorState)=>void) | undefined
  public batteryCallback: ((battery: number)=>void) | undefined

  public get connectionState(): SensorState {
    return this._sensor === undefined ? SensorState.OutOfRange : this._sensor.getState();
  }

  public get batteryPower(): number {
    return this._sensor === undefined ? 0 : this._sensor.getState();
  }

  public get samplingFreq(): SensorSamplingFrequency {
    return this._sensor === undefined ? SensorSamplingFrequency.FrequencyUnsupported : this._sensor.getSamplingFrequency();
  }

  async createAndConnect(info: SensorInfo): Promise<SensorState> {
    return new Promise<SensorState>(async (resolve, reject) => {
      this._scanner?.createSensor(info)
        .then((sensor) => {
          this._sensor = sensor as CallibriSensor

          this._sensor.setSignalTypeCallibri(CallibriSignalType.EMG)
          this._sensor.setSamplingFrequency(SensorSamplingFrequency.FrequencyHz1000)

          this._sensor.setHardwareFilters([SensorFilter.FilterHPFBwhLvl1CutoffFreq1Hz])

          this._sensor.AddConnectionChanged((state) => { 
            if(this.connectionChangedCallback != undefined)
              this.connectionChangedCallback(state); 
          })

          this._sensor.AddBatteryChanged((battery) => { 
            if(this.batteryCallback != undefined)
              this.batteryCallback(battery); 
          })

          if(this.connectionChangedCallback != undefined)
              this.connectionChangedCallback(SensorState.InRange);
          resolve(SensorState.InRange)
        })
        .catch((ex) => { reject(SensorState.OutOfRange) })
    });
  }

  async connectCurrent(): Promise<SensorState> {
    return new Promise<SensorState>(async (resolve, reject) => {
      if(this._sensor?.getState() != SensorState.OutOfRange) {
        resolve(SensorState.InRange)
        return
      }
      this._sensor?.connect()
        .then(() => {
          resolve(SensorState.InRange)
        })
        .catch((ex) => { reject(SensorState.OutOfRange) })
    });
  }

  async disconnectCurrent(): Promise<void> {
    await this._sensor?.disconnect()
  }

  public signalReceivedCallback: ((data: Array<CallibriSignalData>)=>void) | undefined
  public electrodeChangedCallback: ((data: CallibriElectrodeState)=>void) | undefined
  public envelopeReceivedCallback: ((data: Array<CallibriEnvelopeData>)=>void) | undefined

  async startSignal(){
    this._sensor?.AddElectrodeStateChanged((state)=>{
      if(this.electrodeChangedCallback != undefined)
        this.electrodeChangedCallback(state)
    })
    this._sensor?.AddSignalReceived((data)=>{
      if(this.signalReceivedCallback != undefined)
        this.signalReceivedCallback(data)
    })
    await this._sensor?.execute(SensorCommand.StartSignal).catch((ex)=> console.log(ex))
  }

  async stopSignal(){
    this._sensor?.RemoveElectrodeStateChanged()
    this._sensor?.RemoveSignalReceived()
    await this._sensor?.execute(SensorCommand.StopSignal).catch((ex)=> console.log(ex))
  }

  async startEnvelope(){
    this._sensor?.AddEnvelopeDataChanged((data)=>{
      if(this.envelopeReceivedCallback != undefined)
        this.envelopeReceivedCallback(data)
    })
    await this._sensor?.execute(SensorCommand.StartEnvelope).catch((ex)=> console.log(ex))
  }

  async stopEnvelope(){
    this._sensor?.RemoveEnvelopeDataChanged();
    await this._sensor?.execute(SensorCommand.StopEnvelope).catch((ex)=> console.log(ex))
  }

  get info(): string{
    if (this._sensor === undefined || this._sensor.getState() == SensorState.OutOfRange) return `Device unreachable!`;
        var deviceInfo = ``

        var features = this._sensor.getFeatures();
        deviceInfo += `Features:\n`
        features.forEach(feature => {
               deviceInfo += ` ${SensorFeature[feature]}\n`
        });

        var commands = this._sensor.getCommands()
        deviceInfo += `\nCommands:\n`
            commands.forEach(command => {
              deviceInfo += ` ${SensorCommand[command]}\n`
            });

        var parameters = this._sensor.getParameters()
        deviceInfo += '\nParameters:\n'
        parameters.forEach(parameter => { 
          switch(parameter.Param){
            case SensorParameter.Name:
              deviceInfo += ` Name (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getName()}\n`
              break;
            case SensorParameter.State:
              deviceInfo += ` State (${SensorParamAccess[parameter.ParamAccess]}): ${SensorState[this._sensor === undefined ? 1 : this._sensor.getState()]}\n`
              break;
            case SensorParameter.Address:
              deviceInfo += ` Address (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getAddress()}\n`
              break;
            case SensorParameter.SerialNumber:
              deviceInfo += ` Serial number: (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getSerialNumber()}\n`
              break;
            case SensorParameter.FirmwareMode:
              deviceInfo += ` Firmware mode: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorFirmwareMode[this._sensor === undefined ? 1 : this._sensor.getFirmwareMode()]}\n`
              break;
            case SensorParameter.SamplingFrequency:
              deviceInfo += ` Sampling frequency: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorSamplingFrequency[this._sensor === undefined ? 10 : this._sensor.getSamplingFrequency()]}\n`
              break;
            case SensorParameter.Gain:
              deviceInfo += `  Gain: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorGain[this._sensor === undefined ? 11 : this._sensor.getGain()]}\n`
              break;
            case SensorParameter.Offset:
              deviceInfo += ` Offset: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorDataOffset[this._sensor === undefined ? 255 : this._sensor.getDataOffset()]}\n`
              break;
            case SensorParameter.FirmwareVersion:
              var version = this._sensor?.getVersion()
              deviceInfo += ` Firmware version: (${SensorParamAccess[parameter.ParamAccess]}):\n` +
                 `  FW: ${version?.FwMajor}.${version?.FwMinor}.${version?.FwPatch}` +
                 `  HW: ${version?.HwMajor}.${version?.HwMinor}.${version?.HwPatch} ` +
                 `  Ext: ${version?.ExtMajor}\n`
              break;
            case SensorParameter.BattPower:
              deviceInfo += ` Battery power: (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor?.getBattPower()}\n`
              break;
            case SensorParameter.SensorFamily:
              deviceInfo += ` Sensor family: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorFamily[this._sensor === undefined ? 0 : this._sensor.getSensFamily()]}\n`
              break;
            case SensorParameter.ADCInputState:
              deviceInfo += ` ADC input: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorADCInput[this._sensor === undefined ? 0 : this._sensor.getADCInput()]}\n`
              break;
            case SensorParameter.GyroscopeSens:
              deviceInfo += ` Gyro sensitivity: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorGyroscopeSensitivity[this._sensor === undefined ? 0 : this._sensor.getGyroSens()]}\n`
              break;
            case SensorParameter.SamplingFrequencyMEMS:
              deviceInfo += ` Sampling frequency MEMS: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorSamplingFrequency[this._sensor === undefined ? 0 : this._sensor.getSamplingFrequencyMEMS()]}\n`
              break;
            case SensorParameter.AccelerometerSens:
              deviceInfo += ` Accelerometer sensitivity: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorAccelerometerSensitivity[this._sensor === undefined ? 0 : this._sensor.getAccSens()]}\n`
              break;
            case SensorParameter.HardwareFilterState:
              deviceInfo += ` Hardware filters: (${SensorParamAccess[parameter.ParamAccess]}): ${this._sensor === undefined ? 0 : this._sensor.getHardwareFilters()}\n`
              break;
            case SensorParameter.ExternalSwitchState:
              deviceInfo += ` External switch state: (${SensorParamAccess[parameter.ParamAccess]}): ${SensorExternalSwitchInput[this._sensor === undefined ? 0 : this._sensor.getExtSwInput()]}\n`
              break;
          }
        });

        deviceInfo += `\nColor: ${CallibriColorType[this._sensor === undefined ? 4 : this._sensor.getColorCallibri()]}\n`
        deviceInfo += `\Signal type: ${CallibriSignalType[this._sensor === undefined ? 6 : this._sensor.getSignalTypeCallibri()]}\n`

        return deviceInfo
  }
}

const CallibriControllerInstance = CallibriController.Instance;
export default CallibriControllerInstance;
